### PR TITLE
Create dynamic (long press launcher) shortcuts from favorites

### PR DESCRIPTION
**Pull request by morckx**
* also improve the look of the shortcut icons also for pinned shortcuts
* use picasso transforms for circle and rounded-corner homescreen icons
* improve code by using callbacks

https://youtu.be/cX9FYWTpmXQ

### CODE REVIEW

File: app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java [hunk 1]
```diff
@@ -190,4 +207,55 @@ public JSONObject toJson(){
 
 		return null;
 	}
+
+	public interface ShortcutReadyListener {
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java]: I'm inclined to have these two classes outside of `DataRadioStation`.

File: app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java [hunk 2]
```diff
@@ -190,4 +207,55 @@ public JSONObject toJson(){
...
+			super();
+			this.ctx = ctx;
+			this.station = station;
+			this.cb = cb;
+		}
+
+		@Override
+		public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
+			if (Build.VERSION.SDK_INT >= 25) {
+				Intent playByUUUIDintent = new Intent(MediaSessionCallback.ACTION_PLAY_STATION_BY_UUID, null, ctx, ActivityMain.class)
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java]: That's one too many `U`s

File: app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java [hunk 3]
```diff
@@ -190,4 +207,55 @@ public JSONObject toJson(){
...
+						.setIcon(Icon.createWithBitmap(bitmap))
+						.setIntent(playByUUUIDintent)
+						.build();
+				cb.onShortcutReadyListener(shortcut);
+			}
+		}
+
+		@Override
+		public void onBitmapFailed(Drawable errorDrawable) {
+            onBitmapLoaded(((BitmapDrawable)errorDrawable).getBitmap(), null);
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java]: The file probably isn't formatted.

File: app/src/main/java/net/programmierecke/radiodroid2/FavouriteManager.java [hunk 4]
```diff
@@ -23,5 +31,47 @@ public void add(DataRadioStation station){
...
+        ArrayList<ShortcutInfo> shortcuts;
+        int expectedNumber;
+
+        SetDynamicAppLauncherShortcuts(int expectedNumber) {
+            this.expectedNumber = expectedNumber;
+            shortcuts = new ArrayList<ShortcutInfo>(expectedNumber);
+        }
 
+        @Override
+        public void onShortcutReadyListener(ShortcutInfo shortcur) {
```
Comment (werman) [inline app/src/main/java/net/programmierecke/radiodroid2/FavouriteManager.java]: `shortcut`

**Comment by werman** [comment]
@segler-alex Could you give us some window for reviews to be made =) Thanks!

**Comment by segler-alex** [comment]
ok, please always add WIP

File: app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java (hunk 1 above)
Comment (morckx) [reply app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java]: Yes, a shortcut manager class and, even more important, a station icon manager class would make sense.

**Comment by morckx** [comment]
> ok, please always add WIP

OK, I'll use WIP and RFR: ready/request for reviews

File: app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java (hunk 2 above)
Comment (morckx) [reply app/src/main/java/net/programmierecke/radiodroid2/data/DataRadioStation.java]: Thanks, I was already thinking that I need a larger monitor :)

File: app/src/main/java/net/programmierecke/radiodroid2/FavouriteManager.java (hunk 4 above)
Comment (morckx) [reply app/src/main/java/net/programmierecke/radiodroid2/FavouriteManager.java]: I do really need one

**Pull request merged**